Subset composition

warpech opened the pull request:
This PR includes necessary changes for https://github.com/Starcounter/RebelsLounge/issues/179. It works in pair with https://github.com/Starcounter/CompositionProvider/pull/15

The changes are explained in the commit messages.

@alshakero can you pls review?

warpech committed:
Problem: the instructions in CONTRIBUTING.md are wrong

They do not accurately say how to run a development environment on Windows

Solution: adjust the information in CONTRIBUTING.md to what actually works.

warpech committed:
Problem: the ampersand at the end of the npm command causes NodeJS to keep listening even after the console is closed

Solution: this should not be part of the script but just how you use the script. Remove the ampersand from package.json

warpech committed:
Problem: in Firefox and Edge, the <content> tag distributes nodes that should not be distributed (https://github.com/tomalec/Translate-ShadowDOM/issues/3)

Solution: apply the workaround that uses <content select=":not([slot]),[slot='']"></content> instead of <content></content>

warpech committed:
Problem: when Composition$ does not include all partial views, the views not included in Composition$ are not rendered

Solution: for the views not included in Composition$, render the default compositions of these views. This requries additional information (ViewUris) from CompositionProvider. It will be tested in CompositionProvider tests

warpech committed:
Problem: the tests were failing on Travis because I disabled running mockserver in the background

Solution: use "concurrently", which should be equally good to run npm scripts on background in Windows and *nix

alshakero commented:
Can this wait a day? I'm fixing JSONPatcherProxy right now so we don't merge a walkaround. Is this OK? JPP is almost fixed.

alshakero commented:
JSONPatcherProxy PR is ready https://github.com/Palindrom/JSONPatcherProxy/pull/22

warpech committed:
Problem: starcounter-include contained a workaround for an issue that is being fixed in https://github.com/Palindrom/JSONPatcherProxy/pull/22

Solution: remove the workaround

warpech committed:
Problem: starcounter-include incorrectly appended a default composition after the explicit composition was saved as custom composition

Solution: fix the above problem, add tests

alshakero committed:
Problem: Tests were comparing crude HTML Strings

This caused issues in Safari since attributes order
isn't (and doesn't have to be) consistent among browsers.

Solution: Parse HTML strings into nodes, sort attribs alphabetically
then compare each node seperately to decide weather strings are equivalent

alshakero commented:
Here is a reasonable diff (with ignored whitespaces) to my commit https://github.com/Starcounter/starcounter-include/pull/67/commits/912f244814bb86a089dc2955a5a25e2f582245b7?w=1

warpech commented:
Thanks @alshakero. It looks ok to me now. Could you pls merge it, release a version and make a PR to Starcounter?

tomalec commented on starcounter-include.html:
```diff
@@ -219,13 +216,26 @@
+                    if (name === 'starcounter-composition') {
+                        console.warn('`starcounter-composition` was renamed to `declarative-shadow-dom`! Please update your markup, as the old name is likely to be not supported soon (in V1).')
+                    } 
+                    fragment.appendChild(template.content.cloneNode(true));
```
Shouldn't you `importNode` here?

warpech replied:
I don't know, should I? Originally there was `cloneNode` so I kept that: https://github.com/Starcounter/starcounter-include/blob/8725dff915d4bd531ec39f5059c9b4f83e770141/starcounter-include.html#L232

alshakero replied:
Take my comment with a grain of salt, but I thought so too when I was reviewing. But then I figured `imported-templated` imports them for you giving them correct ownership, making cloning sufficient. 

See https://github.com/Juicy/imported-template/blob/1304befa0ff67a43e98f2d644ccb8391535fde59/imported-template.html#L84

tomalec replied:
Thanks @alshakero ! Yep, I think it's not needed as `i-t` imports it.

tomalec commented on starcounter-include.html:
```diff
@@ -238,22 +248,43 @@
+                                        if (this.defaultComposition) {
+                                            this.template.scopedNodes.forEach((nodes) => {
+                                                if(nodes.scope === key) {
+                                                    nodes.forEach((node) => appendComposition(merged, node));
```
Are you appending default composition to the specific composition given in l.266 above?

warpech replied:
Yes

tomalec commented on starcounter-include.html:
```diff
@@ -331,13 +365,12 @@
          */
         StarcounterIncludePrototype.saveLayout = function(compositionStr){
             compositionStr = compositionStr || this.shadowRoot.innerHTML;
-            this.storedLayout = compositionStr;
```
Why don't you update stored layout? as you are right not actually saving/storing new layout.

warpech replied:
I don't know the reason why `storedLayout` exists. I think it is only used by `starcounter-layout-html-editor`. To store the layout, I assign it to `Composition$`.

tomalec replied:
It was there to track: 
- the composition that is saved per given id,
- the one that's actually in DOM - due to layout editing using an editor, **DevTools**, dynamic changes, etc.
- to be able to:
  - revert,
  - see there is something changed, etc.

tomalec commented on starcounter-include.html:
```diff
@@ -114,6 +111,7 @@
          */
         StarcounterIncludePrototype.createComponent = function StarcounterIncludeCreated() {
             var starcounterInclude = this;
+            this.defaultComposition = null;
```
Could you add some comments of `default-` and `explicitComposition` What are those? could you add them to prototype? or at least add both to the constructor, so the API would be there from the beginning?

warpech replied:
- Fallback composition - a single `<slot></slot>` to display all things from light DOM in case no composition was provided
- Default composition - composition found in the view HTML Import (`declarative-shadow-dom` part)
- Custom composition - stored composition provided in JSON by CompositionProvider
- Temporary composition - explicit composition set directly from `starcounter-layout-html-editor`

Should I add this as a comment or to the README?

Yes, I agree:

`StarcounterIncludePrototype.defaultComposition = null;` and `StarcounterIncludePrototype.explicitComposition = null;` should be set.

tomalec replied:
That's great to have solid naming for different means to provide composition, it definitely makes talking about development and testing easier. I think we should for sure have it in JSDocs, and maybe inREADME as well. But not in the way that provides a new definition of a "term" (like formerly partial etc), rather as a clear description how composition could be provided.

tomalec commented on starcounter-include.html:
```diff
@@ -238,22 +248,43 @@
-                            return this.stringToDocumentFragment(this.storedLayout);
+                            const merged = this.stringToDocumentFragment(this.storedLayout);
+
+                            if (compositionProvider.ViewUris) {
```
Could you add some code comments to clarify what's happening here?

warpech replied:
> If there is a composition provided with the new API (`ViewUris`), for all merged views check if this view is included in the composition. If not, this means the composition is a subset that does not contain this view. Search for the default composition for that view and use it instead.

Would this be enough?

tomalec replied:
I'm affraid of using phrase "new API" as it may become old API ;)

tomalec replied:
> If there the composition is provided via sub-/superset API <=> contains `ViewUris`, for all merged views check if this view is included in the composition. If not, this means the composition is a subset that does not contain this view. Search for the default composition for that view and use it instead.

tomalec commented on starcounter-include.html:
```diff
@@ -238,22 +248,43 @@
+                                    const scoped = this.partial[key];
+                                    if (scoped && scoped.Html && compositionProvider.ViewUris.indexOf(scoped.Html) === -1) {
+                                        if (this.defaultComposition) {
+                                            this.template.scopedNodes.forEach((nodes) => {
```
Why are you traversing all the nodes? Wouldn't it be easier to traverse `defaultComposition` it already contains only filtered declarative shadow roots. We could change it to or add a map with a `scope` as a key.

warpech replied:
Yes, perhaps we can do that. Do you want to improve it?

tomalec reviewed (commented):
I still feel a little bit lost in the new structure, I may have more comments once I'll get it.

tomalec commented on starcounter-include.html:
```diff
@@ -238,22 +248,43 @@
+                                            });
+                                        }
+                                        else {
+                                            return false;
```
TL;DR
Shouldn't it me `continue`?

If I get it right, we will end in this `else` if 
- we have specific composition,
- we have some views that are not covered by specific composition, 
- we do not have any default composition.

Shouldn't we return specific composition (`merged`) + maybe `<slot></slot>` to make sure remaining element will get displayed, instead of totally invalidating specific composition for the views that are covered?

warpech replied:
Yeah, I think you are right. Can you fix this and add a test?

tomalec replied:
https://github.com/Starcounter/starcounter-include/issues/72

warpech commented:
I made a new issue to remember about these improvement suggestions https://github.com/Starcounter/starcounter-include/issues/70

tomalec commented:
Sorry, I missed that and create new, duplicate one, slightly more descriptive https://github.com/Starcounter/starcounter-include/issues/72
